Associated functions (#819)
cairo-book/cairo-book: src/ch05-03-method-syntax.md

# Associated functions

> Functions defined within an `impl` block are called _associated functions_ because they have to be associated with one and only one type.

Reviewer: I don't think this is the right approach, not all function inside an `impl` are associated functions, and technically you could write an impl for multiple types.

It's tricky because in Cairo we only have **impls of traits** and not **impl of types** (we have both in rust). So while in Rust, all functions under `impl MyStruct {}` would be associated functions, it's not really the case in Cairo. To bypass that we often use the pattern

```rust
trait MyTypeTrait {
  fn new() -> MyType;
  fn foo(self: MyType)
  fn bar(a: MyType, b: MyType);
}

```

In that case, we're just introducing a useless trait to make something similar to `impl MyType`, hence we often abstract it with the `#[generate_trait]` attribute.

I would keep this part as it was before, perhaps move it to the end of the page, where we can add guidelines and do's / don't regarding trait usage

# Associated functions

> We call _associated functions_ functions that are associated with a `Type`. There are 2 ways for a function to be associated with a `Type`:

Reviewer: ```suggestion
We call _associated functions_ functions that are defined inside an `impl` block and are associated with a specific type. There are 2 ways for a function to be associated with a type `T`:
```

> - By having its first parameter called `self` and of this `Type`.

Reviewer: ```suggestion
- By having its first parameter called `self` and of this type `T`.
```

>   For instance : `area(self :@Rectangle) -> u64` is an associated function to the type `Rectangle`.

Reviewer: ```suggestion
  For instance : `fn area(self :@Rectangle) -> u64` is a function associated to the type `Rectangle`.
```

>   This kind of associated function is called a **method**.

Reviewer: ```suggestion
We call these funtions **methods**, and their first parameter is always `self`, which represents the instance of the type the method is being called on..
```

> - By returning a instance of type `Type` or having at least one parameter of type `Type`.

Reviewer: ```suggestion
- By returning an instance of this type `T`, or having at least one parameter of type `T`.
```

>   For instance : `avg(lhs :@Rectangle, rhs: @Rectangle) -> Rectangle` is an associated function to the type `Rectangle`.

Reviewer: ```suggestion
  For instance, `fn avg(lhs :@Rectangle, rhs: @Rectangle) -> Rectangle` is function associated to the type `Rectangle`.
```

>   This kind of associated function is not a **method** because it hasn't `self` as first parameter.

Reviewer: ```suggestion
  This kind of associated functions don’t have `self` as their first parameter (and thus are not methods) because they don’t need an instance of the type to work with. 
```

{{#include ../listings/ch05-using-structs-to-structure-related-data/no_listing_01_define_methods/src/lib.cairo:trait_implementation}}
```

> Finally, we call this `area` method on the `Rectangle` instance `rect1` using the `<instance_name>.<method_name>` syntax. The instance `rect1` will be passed to the `area` method as the `self` parameter. Methods in Cairo are equivalent to _instance methods_ in object-oriented programming.

Reviewer: ```suggestion
Finally, we call this `area` method on the `Rectangle` instance `rect1` using the `<instance_name>.<method_name>` syntax. The instance `rect1` will be passed to the `area` method as the `self` parameter.
```

Here, we expect that `rect1` can hold `rect2` but not `rect3`.

## Associated functions without `self`

> In Cairo, we can also define a function which doesn't act on a specific instance (so, without any `self` parameter) but which still manipulates the related type and thus is an _associated function_. This is what we call _class methods_ in Object-Oriented programming. As these functions are not called from an instance, we don't use them with the `<instance_name>.<method_name>` syntax but with the `<Trait_or_Impl_name>::<method_name>` syntax as you will see in the next example.

Reviewer: ```suggestion
In Cairo, we can also define a function which doesn't act on a specific instance (so, without any `self` parameter) but which still manipulates the related type. This is what we call _class methods_ in Object-Oriented programming. As these functions are not called from an instance, we don't use them with the `<instance_name>.<method_name>` syntax but with the `<Trait_or_Impl_name>::<method_name>` syntax as you will see in the next example.
```

There’s no strong reason to separate these methods into multiple `trait` and `impl`
blocks here, but this is valid syntax.

## Guidelines about associated functions

> _associated functions_ defined within an `impl` block should all be associated with the same type.

Reviewer: ```suggestion
_Associated functions_ defined within an `impl` block should all be associated with the same type.
```

Here, we expect that `rect1` can hold `rect2` but not `rect3`.

> ## Associated functions without `self` Parameter / Class methods

Reviewer: ```suggestion
## Associated functions
```

I think shorter is good enough

> In Cairo, we can also define a associated function which doesn't act on a specific instance (so, without any `self` parameter) but which still manipulates the related type. This is what we call _class methods_ in Object-Oriented programming. As these methods are not called from an instance, we don't use them with the `<instance_name>.<method_name>` syntax but with the `<Trait_or_Impl_name>::<method_name>` syntax as you will see in the next example.

Reviewer: Actually, the part that you wrote above can be used here as a primer / introductory note to this section, where we define what we call an associated function and how it would be materialized in Cairo, and then keep  the existing text
